### PR TITLE
Language translator updates/fixes

### LINKED ISSUE #493: Text doesn't change when language updated
Spotted by Krzychu: If language is changed in-game, the tutorial (adviser) text doesn't get updated.

Can probably look in to this if we migrate to xml for translations. #336

### PR DESCRIPTION
**Pull request by kvakvs**
Fixes #493

* On TM:PE language change
  * Language in options updates in main menu (before game is loaded) and during the game too
  * Tutorials update 
  * Main TM:PE button tooltip is updated
* All languages loaded at once
* Translation class not static anymore
* Loading CSV from Crowdin instead of multiple language files. The files now moved under `Resources/Translations` in the file tree
* Added en-GB

### CODE REVIEW

**Comment by aubergine10** [comment]
List of things I tried (issues listed later):

- [x] Changing game language 👍 
- [x] Changing lang in mod options prior to loading a save 👍 
- [x] Mod options updates from main menu 👍 
- [x] Changing lang in mod options in-game 👍 
- [x] Mod options updates in-game 👍 
- [x] TM:PE "crown" button: Tooltip ❌ 
- [x] Menu panel: Tooltips on tool buttons 👍 
- [x] Tooltips: Keyboard shortcut modifiers (eg. `Shift`) and special keys (eg. `Backspace`) ❌ 
- [x] Tutorial text (for several tools) ❌ 
- [x] Lane arrows panel 👍 
- [x] Despawn tool: Confirmation dialog ❌ 
- [x] Timed Traffic Lights: Dialog panel 👍 
- [x] Timed Traffic Lights: Overlays ❌ 
- [x] Speed limits: Main panel 👍 
- [x] Speed limits: Default speeds panel 👍 

That's everything I could find that had text to translate (did I miss anything?)...

Issues:

* Hungarian (Magyar) language recently added (#492) isn't included in the language drop-down in mod options...?
* Also, some German translations seem missing:
image
* TM:PE button tool tip (German lang selected) in English:
image
* Should keybinds also be translated (German lang selected):
image image
* Tutorial advisor text did not localise (tried German and Chinese):
image
* Despawn tool confirmation dialog: Yes/No buttons not translated (tried Chinese and German):
image
* Timed traffic lights overlays - I am aware the yellow "Auto" panel is an image and only has few translations so ignore that. However the "CHANGE MODE" panel seems not to translate (as far as I know that is normal text component so should translate):
image

**Comment by aubergine10** [comment]
### Comparing to current `master` branch as requested:

- [x] Changing game language 👍 
- [x] Changing lang in mod options prior to loading a save 👍 
- [x] Mod options updates from main menu 👍 
- [x] Changing lang in mod options in-game 👍 
- [x] Mod options updates in-game 👍 
- [x] TM:PE "crown" button: Tooltip ❌ - same issue as PR
- [x] Menu panel: Tooltips on tool buttons 👍 
- [x] Tooltips: Keyboard shortcut modifiers (eg. `Shift`) and special keys (eg. `Backspace`) ❌ - same issue as PR
- [x] Tutorial text (for several tools) 👍 - this works in `master`
- [x] Lane arrows panel 👍 
- [x] Despawn tool: Confirmation dialog ❌ - same issue as PR
- [x] Timed Traffic Lights: Dialog panel 👍 
- [x] Timed Traffic Lights: Overlays ❌ - same issue as PR
- [x] Speed limits: Main panel ❌ - 'unlimited speed` locale key visible
- [x] Speed limits: Default speeds panel 👍 

So, to summarise, there are two differences between current `master` branch and your PR #509:

1. Tutorial Advisor panels are translating properly (at least when language changed in-game) in the `master` version.
    * Note: This contradicts the issue @krzychu124 was having in #493 
    * In the PR #509 they do not change (always English text no matter the language selection).
2. Noticed the "unlmited speed" caption on speeds panel was showing locale key (tested German language):
image

Note: Regarding point 2 above ("unlimited speed" locale key showing), I might have missed similar issue in the PR build.

**Comment by kvakvs** [comment]
The advisor text update looked like a real bug, unless it is reset somewhere, it would fail adding same keys over again with an exception. One had to remove the existing keys first, and that code was not there in `master`.

**Comment by aubergine10** [comment]
The one thing I've not tested is changing language at main menu, then loading game.

Each time I changed language in main menu, tested mod options from main menu, then reset to English then loaded game, and then started changing language in-game to test UI stuff.

So I'll repeat my tests but this time changing language from main menu to see if I can replicate the issues @krzychu124 was having in `master` branch, and also see if that fixes tutorial text in-game in the PR branch.

**Comment by aubergine10** [comment]
Ok, just replicated the tutorial text issue that @krzychu124 was having.

Steps to reproduce:

1. `master` branch
2. Launch Cities.exe
    * Do NOT alter language from main menu
3. Load a save game
4. Check tutorial - it is in whatever language TM:PE (or game if applicable) was set to when Cities.exe loaded
5. In-game, change language in TM:PE mod options (didn't try game language but assume same issue)
6. Tutorial panels do not change their language

So it seems that, in `master` branch, _changing language from main menu before loading a save_ somehow fixes the bug with tutorial text when later _changing languages in-game_.

If you don't change the language at main menu, then subsequent in-game language changes don't update the tutorial panels.

I will now retest the PR version to see what happens with same steps.

**Comment by aubergine10** [comment]
Just tested the PR again, and got the tutorial text working using same steps as comment above!

1. PR #509 branch
2. Launch Cities.exe
    * Do NOT alter language from main menu
3. Load a save game
4. Check tutorial - it is in whatever language TM:PE (or game if applicable) was set to when Cities.exe loaded
5. In-game, change language in TM:PE mod options (didn't try game language but assume same)
6. Tutorial panels change language properly!

So it seems you've inverted the problem compared to `master` branch 🤣 

The Hungarian (Magyar) language seems to be appearing in mod options again (at least in-game). I will do some testing to see if it's there when I access them via main menu.

I re-tested the "unlimited speed" caption on speeds panel in the PR build - it shows "No limit" English text for some languages (tested German, Chinese, Polish). When I switched to Magyar (Hungarian) it showed correct translation. So I assume we're just missing translations for the other languages.

> Note; We should check the `Speed_limit_unlimited` locale key exists in the lang files. I will do that shortly.

**Comment by aubergine10** [comment]
`Speed_limit_unlimited` is missing from several of the locale text files.

**Comment by kvakvs** [comment]
Ok, until we have determined the translation tool to use, be it Crowdin or something else, or an offline tool, i am not going to address any missing translations. Only those things which are broken by coding or can be fixed by coding.

**Comment by aubergine10** [comment]
Did you manage to sort the issue with the tutorial text?

**Comment by kvakvs** [comment]
The tutorial panel is not refreshed that is why it seems like it doesn't change language, and honestly it shouldn't. If we want it to change language the moment we change TM:PE language, we have to reactivate the tutorial on the same page again. If you click anything in TM:PE menu, the new language is shown.
TM:PE button changes its text, i added an update call for that 4 days ago.
TTL "Light mode" button is a texture, always English and only localized for JP and PL languages.
TTL "Counter" button is a texture, always English and only localized for JP and PL languages.

File: TLM/TLM/UI/Translation.cs [hunk 1]
```diff
@@ -12,77 +17,254 @@ namespace TrafficManager.UI {
...
+                                                   { "Italian", "it" },
+                                                   { "Japanese", "ja" },
+                                                   { "Korean", "ko" },
+                                                   { "Polish", "pl" },
+                                                   { "Portuguese", "pt" },
+                                                   { "Russian", "ru" },
+                                                   { "Spanish", "es" }
+                                               };
+
         private const string DEFAULT_LANGUAGE_CODE = "en";
```
Comment (aubergine10) [inline TLM/TLM/UI/Translation.cs]: Move `const` to top of class?

File: TLM/TLM/UI/Translation.cs [hunk 2]
```diff
@@ -12,77 +17,254 @@ namespace TrafficManager.UI {
     using Util;
 
     public class Translation {
-        public static readonly IDictionary<string, string> LanguageLabels;
-        public static readonly IList<string> AvailableLanguageCodes;
+        /// <summary>
+        /// Defines the order of language selector dropdown.
+        /// </summary>
+        public static readonly IList<string> AvailableLanguageCodes
```
Comment (aubergine10) [inline TLM/TLM/UI/Translation.cs]: Why have separate list of `AvailableLanguageCodes` when we can already get that info from `csv_columns_to_locales_` ?

File: TLM/TLM/UI/Translation.cs [hunk 3]
```diff
@@ -12,77 +17,254 @@ namespace TrafficManager.UI {
...
+                               };
+
+        /// <summary>
+        /// Defines column order in CSV, but not the dropdown order.
+        /// Mapping from translation languages (first row of CSV export) to language locales used in
+        /// <see cref="AvailableLanguageCodes"/> above.
+        /// </summary>
+        private static readonly Dictionary<string, string> csv_columns_to_locales_
+            = new Dictionary<string, string>() {
+                                                   { "English", DEFAULT_LANGUAGE_CODE },
```
Comment (aubergine10) [inline TLM/TLM/UI/Translation.cs]: Replace `DEFAULT_LANGUAGE_CODE` with `en`?

I can see why use it, but it seems a bit strange in this context. I would expect the constant to be used to choose the associated language _from_ the list, rather than being in the list itself.

Also, assuming `en` is going to be Americanised, should its caption become (here and in csv) "English (American)"?

File: TLM/TLM/UI/Translation.cs [hunk 4]
```diff
@@ -12,77 +17,254 @@ namespace TrafficManager.UI {
...
+                                                   { "Chinese Simplified", "zh" },
```
Comment (aubergine10) [inline TLM/TLM/UI/Translation.cs]: We also need a lookup of display language too (what to display in the drop-down in mod options). It looks like this is going to make all the language names in that list English?

Specifically, this sort of thing:

```csharp
            LanguageLabels = new TinyDictionary<string, string> {	
                ["de"] = "Deutsch",	
                ["en"] = "English",	
                ["es"] = "Español",	
                ["fr"] = "Français",	
                ["hu"] = "Magyar",	
                ["it"] = "Italiano",	
                ["ja"] = "日本語",	
                ["ko"] = "한국의",	
                ["nl"] = "Nederlands",	
                ["pl"] = "Polski",	
                ["pt"] = "Português",	
                ["ru"] = "Русский язык",	
                ["zh-tw"] = "中文 (繁體)",	
                ["zh"] = "中文 (简体)"	
            };
```

File: TLM/TLM/UI/Translation.cs (hunk 3 above)
Comment (aubergine10) [reply TLM/TLM/UI/Translation.cs]: We could just put a `// DEFAULT_LANGUAGE_CODE` after that entry.

Also we should maybe arrange it so that both English languages are next to each other.  As the current language is selected by default it shouldn't matter if English isn't top of the list?

File: TLM/TLM/UI/Translation.cs (hunk 1 above)
Comment (kvakvs) [reply TLM/TLM/UI/Translation.cs]: Moved

File: TLM/TLM/UI/Translation.cs (hunk 2 above)
Comment (kvakvs) [reply TLM/TLM/UI/Translation.cs]: The Dictionary does not preserve order. This preserves order. Unless we find some sorting (like A..Z alphabetical will put German first) that looks nice, and then we can rebuild this list at runtime, because Options window will use this array to populate the language selection dropdown.

File: TLM/TLM/UI/Translation.cs (hunk 3 above)
Comment (kvakvs) [reply TLM/TLM/UI/Translation.cs]: We can assume "en" is "en-US" or rename it properly, whatever you prefer. Renaming is a tiny bit more work though.
Language titles are now a string in the translations "I18n Language Name", i might rename it later again, when we figure out some consistent string naming scheme.

File: TLM/TLM/UI/Translation.cs (hunk 4 above)
Comment (kvakvs) [reply TLM/TLM/UI/Translation.cs]: No, this is column names in CSV (language names set by Crowdin go into column names in Excel/CSV)
The actual language names are moved into the translations.

File: TLM/TLM/UI/Translation.cs (hunk 2 above)
Comment (kvakvs) [reply TLM/TLM/UI/Translation.cs]: Replaced with `AvailableLanguageCodes.Values().ToList().Sort()`

**Comment by aubergine10** [comment]
Tested in game and it works great. The pre-existing issues are still there, but IMO we can fix up most of that stuff over time as we overhaul bits of UI.

Can confirm that missing translations now show in English rather than locale keys which is a good enhancement for end users.

Changing languages also seems faster than the old implementation too :)

I don't know if it's possible, but I suspect there are lots of dead strings in the lang file that aren't used any more (eg. stuff from much earlier versions of the mod). Is there a way to identify obsolete strings and/or cull them?

**Comment by kvakvs** [comment]
One way to find dead strings would be to filter all source files for `Translate.GetString` calls, extract those strings and create an Ultimate Map of Every Used Key.
That'd be like 2 commands in Linux shell.
And then finding the intersection can probably be done in Excel. Open the translations CSV, paste the keys we've just extracted, and figure out how to intersect 2 sets of data in Excel.

**Comment by aubergine10** [comment]
I think it's worth doing, as this is ideal point to do it. That way we know we're stating with new format and also a cleaned set of translations.

It would likely also be a good point, seeing all those keys, to do the switch to better key format?

**Comment by kvakvs** [comment]
Switching key format is a tedious task, it needs searching for every key and deciding where it'll go and what name it gets. Decide on format and do at least those from TTL, screaming for a rename

**Comment by kvakvs** [comment]
102 unique strings
```
Translation.GetString("...")
Translation.GetString("Adaptive_step_switching")
Translation.GetString("Add")
Translation.GetString("Add_junction_to_timed_light")
Translation.GetString("Add_step")
Translation.GetString("After_min._time_has_elapsed_switch_to_next_step_if")
Translation.GetString("Allow_all_vehicles")
Translation.GetString("avg._flow")
Translation.GetString("avg._wait")
Translation.GetString("Ban_all_vehicles")
Translation.GetString("Cancel")
Translation.GetString("Clear_Traffic")
Translation.GetString("Copy")
Translation.GetString("default") + ")
Translation.GetString("Default_speed_limit") + ":")
Translation.GetString("Default_speed_limits")
Translation.GetString("Delete")
Translation.GetString("Deselect_all_nodes")
Translation.GetString("Display_speed_limits_mph")
Translation.GetString("down")
Translation.GetString("Driving_to_another_parking_spot")
Translation.GetString("Driving_to_a_parking_spot")
Translation.GetString("Edit")
Translation.GetString("Enable_test_mode_(stay_in_current_step)")
Translation.GetString("Entering_vehicle")
Translation.GetString("Extreme_long_green/red_phases")
Translation.GetString("Extreme_short_green/red_phases")
Translation.GetString("flow_ratio")
Translation.GetString("Hide_counters")
Translation.GetString("High")
Translation.GetString("incoming")
Translation.GetString("Incoming_demand")
Translation.GetString("Invert")
Translation.GetString("Keybind_category_Global")
Translation.GetString("Keybind_category_LaneConnector")
Translation.GetString("Keybind_conflict")
Translation.GetString("Keybind_Exit_subtool")
Translation.GetString("Keybind_lane_connector_delete")
Translation.GetString("Keybind_lane_connector_stay_in_lane")
Translation.GetString("Keybind_toggle_TMPE_main_menu")
Translation.GetString("Keybind_toggle_traffic_lights_tool")
Translation.GetString("Keybind_use_junction_restrictions_tool")
Translation.GetString("Keybind_use_lane_arrow_tool")
Translation.GetString("Keybind_use_lane_connections_tool")
Translation.GetString("Keybind_use_priority_signs_tool")
Translation.GetString("Keybind_use_speed_limits_tool")
Translation.GetString("Kilometers_per_hour")
Translation.GetString("Lane")
Translation.GetString("Lane_Arrow_Changer_Disabled_Connection")
Translation.GetString("Lane_Arrow_Changer_Disabled_Highway")
Translation.GetString("Long_green/red_phases")
Translation.GetString("Looking_for_a_parking_spot")
Translation.GetString("Low")
Translation.GetString("Max._Time:")
Translation.GetString("Miles_per_hour")
Translation.GetString("min/max")
Translation.GetString("Min._Time:")
Translation.GetString("Moderate_green/red_phases")
Translation.GetString("Node_is_level_crossing")
Translation.GetString("NODE_IS_TIMED_LIGHT")
Translation.GetString("Node") + " " + t + "\n")
Translation.GetString("Outgoing_demand")
Translation.GetString("Paste")
Translation.GetString("Remove_junction_from_timed_light")
Translation.GetString("Remove_this_citizen")
Translation.GetString("Remove_this_vehicle")
Translation.GetString("Remove_timed_traffic_light")
Translation.GetString("Road_type") + ":")
Translation.GetString("Rotate_left")
Translation.GetString("Rotate_right")
Translation.GetString("Save")
Translation.GetString("Save") + " & " + Translation.GetString("Apply")
Translation.GetString("Segment")
Translation.GetString("Select_junction")
Translation.GetString("Select_nodes")
Translation.GetString("Select_nodes_windowTitle")
Translation.GetString("Sensitivity")
Translation.GetString("Setup_timed_traffic_light")
Translation.GetString("Short_green/red_phases")
Translation.GetString("Show_counters")
Translation.GetString("Show_lane-wise_speed_limits")
Translation.GetString("Skip")
Translation.GetString("Speed_limits")
Translation.GetString("Speed_limit_unlimited")
Translation.GetString("Start")
Translation.GetString("State")
Translation.GetString("Stop")
Translation.GetString("Switch_view")
Translation.GetString("Thinking_of_a_good_parking_spot")
Translation.GetString("Timed_traffic_lights_manager")
Translation.GetString("Traffic_Manager_detected_incompatible_mods")
Translation.GetString("Unsubscribe")
Translation.GetString("up")
Translation.GetString("Using_public_transport")
Translation.GetString("Vehicle_restrictions")
Translation.GetString("Vehicles_may_enter_blocked_junctions")
Translation.GetString("Very_long_green/red_phases")
Translation.GetString("Very_short_green/red_phases")
Translation.GetString("View")
Translation.GetString("wait_ratio")
Translation.GetString("Walking")
Translation.GetString("Walking_to_car")
```

**Comment by aubergine10** [comment]
Just jotting this down here so I don't forget:

* Length of string for a checkbox item in mod options: 70 chars max
* Indented checkbox item: 60 chars max

I've been adding those to the context for translations but might have missed some :)

Also added some screen shots - the text recognition tool is freaking awesome: Just drag box over text on screenshot and it does OCR and links it to the translation (shows list if more than one match, so you can select the right one). ❤️

**Comment by aubergine10** [comment]
I think we need to be consistent with prefixes (eg. `Button.` or `Checkbox.` etc).

Either everything get a prefix, or nothing gets a prefix. I suggest the former (everything given prefix).

For example, there is extremely limited space for tabs in mod options, but the same string is sometimes used as a section heading as well (which can be much longer).

So even though it's same text, we should have `Tab.Whatever` and `Heading.Whatever`.

I think we should do this globally for all strings, even though that is a giant grindy task, because once this goes live it will be a freaking nightmare to change from that point onwards.

**Comment by aubergine10** [comment]
I think we should add a separate `InfoViews.csv` even though there is very little there at the moment.

There are a few strings currently residing in `Menu.csv` that would get moved across:

* `Incoming demand`
* `Outgoing demand`
* `Button.Switch view`

**Comment by aubergine10** [comment]
The following `ModConflicts.csv` strings should be moved to `Options.csv`:

* `Checkbox.Scan for known incompatible mods on startup`
* `Ignore disabled mods`

**Comment by aubergine10** [comment]
The following `SpeedLimits.csv` strings should be moved to `Options.csv`:

* `Theme.*`
* `Road signs theme mph`

**Comment by kvakvs** [comment]
OK
let's try establish some rules then
* `Tooltip.` for pop up tooltips
* `Label.` for UI labels, most likely length is limited
* `Group.` for UI groups, same limits as labels
* `Tab.` for UI tabs
- [ ] `Checkbox.`  for checkboxes, length is limited sometimes
* `Window.` for window titles
* `Error.`  and `Warning.` for multiline messages inside message boxes
* `Format.` for format strings

For example to group things further, in options file i prefixed strings with their tab name: `Overlay.Buildings` means a buildings label in the Overlay tab.

**Comment by aubergine10** [comment]
For overlays, maybe something like:

* `Overlays.Label:Whatever`
* `Maintenance.Button:Whatever`

So all `Overlays` will be grouped together, then all component types will be the secondary grouping.

For dialog boxes (eg. confirm despawn, or the mod compatibility checker) what about:

* `Dialog.Text:...` - body text of dialog (eg. for confirm dialogs)
* `Dialog.Title:...` - titlebar text
* `Dialog.Button:...` etc.?

**Comment by aubergine10** [comment]
Couple of buttons in `Menu.csv` belong elsewhere:

- [ ] `Button.Remove this citizen` -> `AI_Citizen.csv`
- [ ] `Button.Remove this vehicle` -> `AI_Car.csv`

**Comment by kvakvs** [comment]
@aubergine10 the stuff's ready for testing
Except a few strings which could be moves elsewhere I am quite happy with it.

**Comment by aubergine10** [comment]
Looking good from quick glance. I need to do some more testing and will report back any issues I find.

**Comment by kvakvs** [comment]
what are the outstanding issues? I am not fixing pre-existing issues. We've fixed the string names and missing translations issues. And some testing was done too. What we do with it now? Can we deliver or fix and deliver?
I've noticed some new translations been suggested, we can approve these and I can update the strings (easy download and unzip in Resources/Translations).

**Comment by aubergine10** [comment]
I might have botched some merge conflicts on my side, but noticed the issues below - can anyone else confirm?

Lane connector button tooltip:
image

Toggle despawn button tool tip:
image

Junction restrictions tool tip:
image

Gameplay options:
image

Policies options:
image

Overlays options:
image

Maintenance tab:
image

Everything else I checked looked OK (in English at least).

**Comment by kvakvs** [comment]
These can be discovered visually by looking into exported Excel tables. English column should never be empty, because it will be shown if the other language column is empty. If both other language and English columns were empty, you will get the key back, sometimes even with that ¶ (pilcrow) sign.

**Comment by aubergine10** [comment]
Fixed issue with junction restrictions tooltip (was missing `Tooltip:` prefix on translation site) and updated csv files to latest build which seems to have resolved all the other issues.

**Comment by aubergine10** [comment]
Docs updated in wiki: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Localisation

**Comment by kvakvs** [comment]
In ModConflicts "Checkbox:Ignore disabled mods" has the key copied in some of the languages. I am not sure why this is happening, we've done multiple transforms of the translations on the way like exporting, reimporting, moving and renaming strings, so probably there's a bug in Crowdin somewhere which we could report. Anyway, the key is copied in some languages there and those columns should be cleared, or the string possibly could be restored from the legacy strings files.

**Comment by kvakvs** [comment]
String "Toolip:Toggle main menu" has a typo in the key.

**Comment by aubergine10** [comment]
They are easy to spot so I'll just edit the csv files

**Comment by aubergine10** [comment]
Looks like changing options on crowdin solved it - just rebuilt and the duplicates seem to be removed.

Pushed a commit with latest versions, also removed old lang files.

Let me know if there are any remaining issues.

**Comment by kvakvs** [comment]
I am very happy with how it turned out.
Wait for krzychu124 ? Or merge away? :)

**Comment by aubergine10** [comment]
I think merge, it's had lots of testing.

**Comment by krzychu124** [comment]
Yeah, I've tested it yesterday and surprisingly I didn't spotted any issue 😉
